src/main/java/io/cdap/plugin/google/shopping/content/source/ShoppingContentClient.java

```diff
@@ -0,0 +1,159 @@
+package io.cdap.plugin.google.shopping.content.source;
+
+import com.google.api.client.auth.oauth2.Credential;
+import com.google.api.client.googleapis.auth.oauth2.GoogleCredential;
+import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport;
+import com.google.api.client.http.HttpRequestInitializer;
+import com.google.api.client.http.HttpTransport;
+import com.google.api.client.json.jackson2.JacksonFactory;
+import com.google.api.services.content.ShoppingContent;
+import com.google.api.services.content.ShoppingContentScopes;
+import com.google.api.services.content.model.Product;
+import com.google.api.services.content.model.ProductsListResponse;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.math.BigInteger;
+import java.security.GeneralSecurityException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Client to read from Google Shopping Content API.
+ * It provides an iterator interface to get Products from Shopping Content API.
+ */
+public class ShoppingContentClient {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(ShoppingContentClient.class);
+
+  private static final long MAX_PRODUCTS = 10L;
+
+  private final String serviceAccountPath;
+
+  private final BigInteger merchantId;
+
+  private final ShoppingContent shoppingContent;
+
+  private final HttpTransport httpTransport;
+
+  private String nextPageToken = null;
+
+  private boolean end = false;
+
+  private List<Product> currentPage;
+
+  public ShoppingContentClient(String serviceAccountPath,
+      String merchantId) throws IOException, GeneralSecurityException {
+    this.serviceAccountPath = serviceAccountPath;
+    this.merchantId = new BigInteger(merchantId);
+    this.httpTransport = createHttpTransport();
+
+    HttpRequestInitializer initializer;
+
+    if (serviceAccountPath.isEmpty()) {
+      // Uses default credential is service account file is not provided.
+      initializer = GoogleCredential.getApplicationDefault();
+    } else {
+      initializer = loadCredential(serviceAccountPath, httpTransport);
```

Comment:
it's usually not a good idea to create resources in a constructor. It would be cleaner for the constructor to simply set values, and have an initialize() method that need to be called before the client could be used.

It also seems like the contents of this class could be moved to the RecordReader, which the ShoppingContent service built in the initialize() method there. Unless there are plans for something else to use this client?

Reply:
thx, updated accordingly, but I kept ShoppingContentClient class there, as I think this makes RecordReader and ShoppingContentSplit more modular?